Fix timers running inside scene preview on drag

lightspot21 opened the pull request:
This PR closes #39221.

~~As a hacky work-around, we iterate through all first-level children of the scene's root and call `stop()` on the ones that are Timers.~~

~~This approach is bad, and I'm currently thinking of using the pause function of the scene tree, but for some reason Godot crashes when toggling pause, which I'll need to investigate.~~

New approach is to pause the subtree which starts with the node containing the preview, instead of iterating through all of its children. This ensures that the preview stays frozen regardless of the existence of second-level timers (unless the child nodes are set to `Process` pause mode) and is cleaner than the old one, IMO.

(This is not a duplicate PR, the old one was not based on the `master` branch due to an oversight of mine)

EDIT: Waiting for the `OS/DisplayServer` split to be completed ~~(or at least previews working again on master)~~
Previews are working again, so this is ready for review.

lightspot21 committed:
Stop all first-level Timer children of a Scene on preview

volzhs commented:
pausing scene preview is not good I think.
it is helpful when placing particle scene or etc.

lightspot21 commented:
@volzhs I don't know if this is intended behavior, paging @Calinou

volzhs commented:
I just set up test scene to demonstrate it.
![instance_scene](https://user-images.githubusercontent.com/8281454/83529921-fae66e80-a525-11ea-9e05-7ce54ad279e4.gif)
it is pretty helpful with viewing live update instanced scene.

lightspot21 commented:
Do we want this as a feature or should I fix it? @Calinou @KoBeWi

lightspot21 committed:
Fix clang-format issue and convert TODO to a #warning

The comment lines inside the file were quite unsightly and broke
formatting rules. To amend this, the TODO is now a warning at
compile time, while the lines describing why the method is a hack
are presented below:

This fix works if the timer on the scene is an immediate child
(it breaks on nested timers). I tried to use something like
"instance->get_tree()->set_pause(true)"
to pause updates on the whole scene tree, but Godot crashes, so more
investigation is needed.

KoBeWi commented:
The fix itself isn't really complicated, so unless the issue is really unimportant, it should be ok to have this (but I wonder if there are other nodes that could be taken care of too).

KoBeWi commented on editor/plugins/canvas_item_editor_plugin.cpp:
```diff
@@ -5996,7 +5996,16 @@ void CanvasItemEditorViewport::_create_preview(const Vector<String> &files) cons
 					if (instance) {
+						for (int j = 0; j < instance->get_child_count(); j++) {
+							Node *curr_child = instance->get_child(j);
+							if (curr_child->get_class() == "Timer") {
```
The class checks in source are usually done like this:
```
Timer *timer = Object::cast_to<Timer>(instance->get_child(j));
if (timer)
```
This also allows you to call `stop()` directly.

lightspot21 replied:
I'll keep it in mind, thank you!
What I don't currently like with the way I handled this is that it works only for timers that are immediate children of the scene's root node. IMO, a better way would be to set the scene's pause state (by using something like `instance->get_tree()->set_pause(true)`) but as I stated on the issue Godot crashes when calling this, and I need to find out why.

KoBeWi replied:
You could also recursively stop the Timers, but just sounds like overkill. IMO in most cases stopping immediate Timers should be enough, the issue is not common anyways.

lightspot21 commented:
> The fix itself isn't really complicated, so unless the issue is really unimportant, it should be ok to have this (but I wonder if there are other nodes that could be taken care of too).

I didn't quite get what you mean here: by "this" you mean to have the behavior after the fix, or the behavior as it is now?

KoBeWi commented:
I meant the fix.

lightspot21 committed:
Change class check according to review

lightspot21 committed:
Replace for-loop with pausing of the preview node

The old method of looping through all children and checking their type
to determine if they are Timers was kind of crude, so this commit replaces
it with a better one, namely pausing the whole subtree of the preview node
at once.

KoBeWi reviewed (commented):
I looked at the newest changes and it looks like something that would go well with #39606.

Setting the pause mode of the editor is very hacky, especially in such a random place. If anything, it could be done in the constructor. But IMO your first solution of simply iterating over Timers and stopping them would be enough, although I'd like others to comment on this too .-.

Alternatively, you could wait until #39606 is merged. Then pausing the preview selectively could be the best fix.

akien-mga commented:
Sorry for the long delay reviewing this further. The current implementation seems OK from a cursory look, though I'm not convinced if the actual "bug" being fixed is something that needs to be fixed in the first place - but I don't have a strong opinion either way.

If @KoBeWi approves the implementation this can likely be rebased to squash the commits and solve merge conflicts, and merged.

KoBeWi commented on editor/plugins/canvas_item_editor_plugin.cpp:
```diff
@@ -6039,6 +6039,9 @@ void CanvasItemEditorViewport::_create_preview(const Vector<String> &files) cons
 					Node *instance = scene->instance();
 					if (instance) {
 						preview_node->add_child(instance);
+						preview_node->set_pause_mode(PAUSE_MODE_STOP);
```
After the pause rework, you no longer need to pause the tree, just set `process_mode` of the `preview_node` to `PROCESS_NODE_DISABLED`. It will only stop the preview node, making the solution a one-liner.

KoBeWi commented:
tbh I'm not sure myself if it's worth fixing; I opened the issue because the behavior was unexpected to me. I just checked and freeing the preview node doesn't cause crash, so the issue is rather harmless.

akien-mga commented:
I'd say let's shelve it for now, and revisit if we find it to actually be a problem worth fixing.
